# 1. Bronze to silver transformation 

Note: on orginal code for start session 

In [ ]:
# from delta import *
# from pyspark.sql import SparkSession

# builder = SparkSession.builder \
#     .appName("RetailPipeline") \
#     .config("spark.sql.extensions", "io.delta.sql.DeltaSparkSessionExtension") \
#     .config("spark.sql.catalog.spark_catalog", "org.apache.spark.sql.delta.catalog.DeltaCatalog")

# spark = configure_spark_with_delta_pip(builder).getOrCreate()

## Section 1: read

# this is markdown

In [3]:
try:
    spark
except NameError:
    from pyspark.sql import SparkSession
    from delta import configure_spark_with_delta_pip

    builder = SparkSession.builder \
        .appName("retail_pipeline") \
        .master("local[*]") \
        .config("spark.sql.extensions", "io.delta.sql.DeltaSparkSessionExtension") \
        .config("spark.sql.catalog.spark_catalog", "org.apache.spark.sql.delta.catalog.DeltaCatalog")
    
    spark = configure_spark_with_delta_pip(builder).getOrCreate()

ERROR StatusLogger Reconfiguration failed: No configuration found for '5c29bfd' at 'null' in 'null'
ERROR StatusLogger Reconfiguration failed: No configuration found for 'Default' at 'null' in 'null'
26/04/24 13:20:23 WARN Utils: Your hostname, YTuSurface5 resolves to a loopback address: 127.0.1.1; using 10.255.255.254 instead (on interface lo)
26/04/24 13:20:23 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address


:: loading settings :: url = jar:file:/home/yiqingtu/spark/spark-3.5.5-bin-hadoop3/jars/ivy-2.5.1.jar!/org/apache/ivy/core/settings/ivysettings.xml


Ivy Default Cache set to: /home/yiqingtu/.ivy2/cache
The jars for the packages stored in: /home/yiqingtu/.ivy2/jars
io.delta#delta-spark_2.12 added as a dependency
:: resolving dependencies :: org.apache.spark#spark-submit-parent-e4244e10-cd31-464e-9345-f82e0d686edd;1.0
	confs: [default]
	found io.delta#delta-spark_2.12;3.2.0 in central
	found io.delta#delta-storage;3.2.0 in central
	found org.antlr#antlr4-runtime;4.9.3 in central
:: resolution report :: resolve 249ms :: artifacts dl 23ms
	:: modules in use:
	io.delta#delta-spark_2.12;3.2.0 from central in [default]
	io.delta#delta-storage;3.2.0 from central in [default]
	org.antlr#antlr4-runtime;4.9.3 from central in [default]
	---------------------------------------------------------------------
	|                  |            modules            ||   artifacts   |
	|       conf       | number| search|dwnlded|evicted|| number|dwnlded|
	---------------------------------------------------------------------
	|      default     |   3   |

In [26]:
folder = "../delta/"

In [45]:
path = folder + "bronze/orders"
df_bronze = spark.read.load(path)

# need panda library
df_bronze.limit(1000).toPandas()

,row_id,order_id,order_date,ship_date,ship_mode,customer_id,customer_name,segment,country,city,...,postal_code,region,product_id,category,subcategory,product_name,sales,quantity,discount,profit
0,764,CA-2014-162775,2014-01-13,2014-01-15,Second Class,CS-12250,Chris Selesnick,Corporate,United States,Bossier City,...,71111,South,OFF-EN-10001990,Office Supplies,Envelopes,Staple envelope,11.36,2.0,0.0,5.3392
1,765,CA-2014-162775,2014-01-13,2014-01-15,Second Class,CS-12250,Chris Selesnick,Corporate,United States,Bossier City,...,71111,South,OFF-EN-10001532,Office Supplies,Envelopes,Brown Kraft Recycled Envelopes,50.94,3.0,0.0,25.4700
2,766,CA-2014-162775,2014-01-13,2014-01-15,Second Class,CS-12250,Chris Selesnick,Corporate,United States,Bossier City,...,71111,South,TEC-AC-10003174,Technology,Accessories,Plantronics S12 Corded Telephone Headset System,646.74,6.0,0.0,258.6960
3,767,CA-2014-162775,2014-01-13,2014-01-15,Second Class,CS-12250,Chris Selesnick,Corporate,United States,Bossier City,...,71111,South,OFF-BI-10004187,Office Supplies,Binders,3-ring staple pack,5.64,3.0,0.0,2.7072
4,768,CA-2014-162775,2014-01-13,2014-01-15,Second Class,CS-12250,Chris Selesnick,Corporate,United States,Bossier City,...,71111,South,OFF-ST-10000025,Office Supplies,Storage,Fellowes Stor/Drawer Steel Plus Storage Drawers,572.58,6.0,0.0,34.3548
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
94,6475,CA-2014-149524,2014-01-14,2014-01-15,First Class,BS-11590,Brendan Sweed,Corporate,United States,Philadelphia,...,19140,East,FUR-BO-10003433,Furniture,Bookcases,Sauder Cornerstone Collection Library,61.96,4.0,0.5,-53.2856
95,6475,CA-2014-149524,2014-01-14,2014-01-15,First Class,BS-11590,Brendan Sweed,Corporate,United States,Philadelphia,...,19140,East,FUR-BO-10003433,Furniture,Bookcases,Sauder Cornerstone Collection Library,61.96,4.0,0.5,-53.2856
96,717,CA-2014-130092,2014-01-11,2014-01-14,First Class,SV-20365,Seth Vernon,Consumer,United States,Dover,...,19901,East,FUR-FU-10000010,Furniture,Furnishings,"DAX Value U-Channel Document Frames, Easel Back",9.94,2.0,0.0,3.0814
97,717,CA-2014-130092,2014-01-11,2014-01-14,First Class,SV-20365,Seth Vernon,Consumer,United States,Dover,...,19901,East,FUR-FU-10000010,Furniture,Furnishings,"DAX Value U-Channel Document Frames, Easel Back",9.94,2.0,0.0,3.0814


In [46]:
# check table
df_check = spark.read.load(path)

print("Rows:", df_check.count())
df_check.printSchema()
df_bronze.show(5, truncate=False)

Rows: 99
root
 |-- row_id: integer (nullable = true)
 |-- order_id: string (nullable = true)
 |-- order_date: date (nullable = true)
 |-- ship_date: date (nullable = true)
 |-- ship_mode: string (nullable = true)
 |-- customer_id: string (nullable = true)
 |-- customer_name: string (nullable = true)
 |-- segment: string (nullable = true)
 |-- country: string (nullable = true)
 |-- city: string (nullable = true)
 |-- state: string (nullable = true)
 |-- postal_code: integer (nullable = true)
 |-- region: string (nullable = true)
 |-- product_id: string (nullable = true)
 |-- category: string (nullable = true)
 |-- subcategory: string (nullable = true)
 |-- product_name: string (nullable = true)
 |-- sales: string (nullable = true)
 |-- quantity: double (nullable = true)
 |-- discount: double (nullable = true)
 |-- profit: double (nullable = true)

+------+--------------+----------+----------+------------+-----------+---------------+---------+-------------+------------+---------+--------

## section2: clean data


In [6]:
from pyspark.sql.functions import col
df_bronze = df_bronze \
    .withColumn("sales", col("sales").cast("double")) \
    .withColumn("quantity", col("quantity").cast("int"))

In [7]:
df_bronze.filter(
    col("sales").isNull() | col("profit").isNull()
).show()

+------+--------------+----------+----------+--------------+-----------+-------------+--------+-------------+-------+-----+-----------+-------+---------------+---------------+-----------+--------------------+-----+--------+--------+------+
|row_id|      order_id|order_date| ship_date|     ship_mode|customer_id|customer_name| segment|      country|   city|state|postal_code| region|     product_id|       category|subcategory|        product_name|sales|quantity|discount|profit|
+------+--------------+----------+----------+--------------+-----------+-------------+--------+-------------+-------+-----+-----------+-------+---------------+---------------+-----------+--------------------+-----+--------+--------+------+
|  7981|CA-2014-103800|2014-01-03|2014-01-07|Standard Class|   DP-13000|Darren Powers|Consumer|United States|Houston|Texas|      77095|Central|OFF-PA-10000174|Office Supplies|      Paper|"Message Book, Wi...| NULL|      16|     2.0|   0.2|
|  7981|CA-2014-103800|2014-01-03|2014-0

In [8]:
# how many null

df_bronze.filter(col("sales").isNull()).count()

3

In [9]:
# drop null and repeat records

# For business: customer_id is required for downstream customer analytics
df_bronze = df_bronze.filter(col("sales").isNotNull())


## section3: deduplicate


In [10]:
# step 2 remove duplicate

from pyspark.sql.functions import count

df_bronze.groupBy("order_id") \
    .count() \
    .filter("count > 1") \
    .show()


+--------------+-----+
|      order_id|count|
+--------------+-----+
|CA-2014-118192|    6|
|CA-2014-109232|    3|
|CA-2014-162775|   15|
|CA-2014-157147|    9|
|CA-2014-130813|    3|
|CA-2014-106054|    3|
|CA-2014-167199|   21|
|CA-2014-105417|    6|
|CA-2014-112326|    9|
|CA-2014-149020|    6|
|CA-2014-135405|    6|
|CA-2014-141817|    3|
|CA-2014-149524|    3|
|CA-2014-130092|    3|
+--------------+-----+



In [11]:
df_bronze.count() - df_bronze.dropDuplicates().count()

64

In [12]:
total = df_bronze.count()
distinct = df_bronze.distinct().count()

print("Duplicate rows:", total - distinct)

Duplicate rows: 64


### row level duplicates

In [13]:
df_bronze.groupBy(df_bronze.columns) \
    .count() \
    .filter("count > 1") \
    .show(100)

+------+--------------+----------+----------+--------------+-----------+-----------------+-----------+-------------+--------------+--------------+-----------+-------+---------------+---------------+-----------+--------------------+-------+--------+--------+--------+-----+
|row_id|      order_id|order_date| ship_date|     ship_mode|customer_id|    customer_name|    segment|      country|          city|         state|postal_code| region|     product_id|       category|subcategory|        product_name|  sales|quantity|discount|  profit|count|
+------+--------------+----------+----------+--------------+-----------+-----------------+-----------+-------------+--------------+--------------+-----------+-------+---------------+---------------+-----------+--------------------+-------+--------+--------+--------+-----+
|  9629|CA-2014-118192|2014-01-13|2014-01-18|Standard Class|   MM-17920|    Michael Moore|   Consumer|United States|        Newark|          Ohio|      43055|   East|OFF-PA-10002947

In [14]:
# produce report
row_dupes_df = df_bronze.groupBy(df_bronze.columns) \
    .count() \
    .filter(col("count") > 1)


In [15]:
row_duplicates = df_bronze.join(
    row_dupes_df.drop("count"),
    on=df_bronze.columns,
    how="inner"
)

In [40]:

path = folder +  "quanrantine/row_level_duplicates"
row_duplicates.write.format("delta") \
    .mode("overwrite") \
    .save(folder + path)

26/04/24 13:51:22 WARN MemoryManager: Total allocation exceeds 95.00% (1,020,054,720 bytes) of heap memory
Scaling row group sizes to 95.00% for 8 writers
26/04/24 13:51:22 WARN MemoryManager: Total allocation exceeds 95.00% (1,020,054,720 bytes) of heap memory
Scaling row group sizes to 84.44% for 9 writers
26/04/24 13:51:22 WARN MemoryManager: Total allocation exceeds 95.00% (1,020,054,720 bytes) of heap memory
Scaling row group sizes to 95.00% for 8 writers


In [41]:
# check table
df_check = spark.read.load(path)

print("Rows:", df_check.count())
df_check.printSchema()
df_bronze.show(5, truncate=False)

Rows: 96
root
 |-- row_id: integer (nullable = true)
 |-- order_id: string (nullable = true)
 |-- order_date: date (nullable = true)
 |-- ship_date: date (nullable = true)
 |-- ship_mode: string (nullable = true)
 |-- customer_id: string (nullable = true)
 |-- customer_name: string (nullable = true)
 |-- segment: string (nullable = true)
 |-- country: string (nullable = true)
 |-- city: string (nullable = true)
 |-- state: string (nullable = true)
 |-- postal_code: integer (nullable = true)
 |-- region: string (nullable = true)
 |-- product_id: string (nullable = true)
 |-- category: string (nullable = true)
 |-- subcategory: string (nullable = true)
 |-- product_name: string (nullable = true)
 |-- sales: double (nullable = true)
 |-- quantity: integer (nullable = true)
 |-- discount: double (nullable = true)
 |-- profit: double (nullable = true)

+------+--------------+----------+----------+------------+-----------+---------------+---------+-------------+------------+---------+-------

### Key level duplicates 

# decect 
from pyspark.sql.functions import col, count

key_dupes_df = df_bronze.groupBy("order_id") \
    .count() \
    .filter(col("count") > 1)

# build report
total_dup_keys = key_dupes_df.count()

total_dup_rows = df_bronze.join(
    key_dupes_df.select("order_id"),
    on="order_id",
    how="inner"
).count()

print("Duplicate order_ids:", total_dup_keys)
print("Total duplicate rows:", total_dup_rows)

In [29]:
from pyspark.sql.functions import col, count

key_dupes_df = df_bronze.groupBy("order_id") \
    .count() \
    .filter(col("count") > 1)

In [30]:
total_dup_keys = key_dupes_df.count()

total_dup_rows = df_bronze.join(
    key_dupes_df.select("order_id"),
    on="order_id",
    how="inner"
).count()

print("Duplicate order_ids:", total_dup_keys)
print("Total duplicate rows:", total_dup_rows)

Duplicate order_ids: 14
Total duplicate rows: 96


In [37]:
key_duplicates = df_bronze.join(
    key_dupes_df.select("order_id"),
    on="order_id",
    how="inner"
)
path = folder + "quarantine/key_duplicates"
key_duplicates.write.format("delta") \
    .mode("overwrite") \
    .save(path)

# save duplicates
path = folder + "quarantine/key_duplicates"

key_duplicates = df_bronze.join(
    key_dupes_df.select("order_id"),
    on="order_id",
    how="inner"
)

key_duplicates.write.format("delta") \
    .mode("overwrite") \
    .save(path)
   
   

In [39]:
# check table
df_check = spark.read.load(path)

print("Rows:", df_check.count())
df_check.printSchema()
df_bronze.show(5, truncate=False)


Rows: 288
root
 |-- order_id: string (nullable = true)
 |-- row_id: integer (nullable = true)
 |-- order_date: date (nullable = true)
 |-- ship_date: date (nullable = true)
 |-- ship_mode: string (nullable = true)
 |-- customer_id: string (nullable = true)
 |-- customer_name: string (nullable = true)
 |-- segment: string (nullable = true)
 |-- country: string (nullable = true)
 |-- city: string (nullable = true)
 |-- state: string (nullable = true)
 |-- postal_code: integer (nullable = true)
 |-- region: string (nullable = true)
 |-- product_id: string (nullable = true)
 |-- category: string (nullable = true)
 |-- subcategory: string (nullable = true)
 |-- product_name: string (nullable = true)
 |-- sales: double (nullable = true)
 |-- quantity: integer (nullable = true)
 |-- discount: double (nullable = true)
 |-- profit: double (nullable = true)
 |-- order_year: integer (nullable = true)

+------+--------------+----------+----------+------------+-----------+---------------+---------+

## section 4: check business rule

In [19]:
# Business check: sales must be >= 0 because refunds are stored separately
df_bronze.filter("sales <= 0").count()

0

## section 5: write silver

## Section 5: Write to silver table

In [42]:
print("Rows:", df_bronze.count())
df_bronze.printSchema()
df_bronze.show(5, truncate=False)

Rows: 96
root
 |-- row_id: integer (nullable = true)
 |-- order_id: string (nullable = true)
 |-- order_date: date (nullable = true)
 |-- ship_date: date (nullable = true)
 |-- ship_mode: string (nullable = true)
 |-- customer_id: string (nullable = true)
 |-- customer_name: string (nullable = true)
 |-- segment: string (nullable = true)
 |-- country: string (nullable = true)
 |-- city: string (nullable = true)
 |-- state: string (nullable = true)
 |-- postal_code: integer (nullable = true)
 |-- region: string (nullable = true)
 |-- product_id: string (nullable = true)
 |-- category: string (nullable = true)
 |-- subcategory: string (nullable = true)
 |-- product_name: string (nullable = true)
 |-- sales: double (nullable = true)
 |-- quantity: integer (nullable = true)
 |-- discount: double (nullable = true)
 |-- profit: double (nullable = true)
 |-- order_year: integer (nullable = true)

+------+--------------+----------+----------+------------+-----------+---------------+---------+-

In [43]:
from pyspark.sql.functions import year
silver_path = folder + "silver/orders"

df_bronze = df_bronze.withColumn("order_year", year(col("order_date")))

df_bronze.write.format("delta") \
    .mode("overwrite") \
    .partitionBy("order_year") \
    .save(silver_path)

In [44]:
# check table written

df_check = spark.read.load(silver_path)

print("Rows:", df_check.count())
df_check.printSchema()
df_bronze.show(5, truncate=False)

Rows: 96
root
 |-- row_id: integer (nullable = true)
 |-- order_id: string (nullable = true)
 |-- order_date: date (nullable = true)
 |-- ship_date: date (nullable = true)
 |-- ship_mode: string (nullable = true)
 |-- customer_id: string (nullable = true)
 |-- customer_name: string (nullable = true)
 |-- segment: string (nullable = true)
 |-- country: string (nullable = true)
 |-- city: string (nullable = true)
 |-- state: string (nullable = true)
 |-- postal_code: integer (nullable = true)
 |-- region: string (nullable = true)
 |-- product_id: string (nullable = true)
 |-- category: string (nullable = true)
 |-- subcategory: string (nullable = true)
 |-- product_name: string (nullable = true)
 |-- sales: double (nullable = true)
 |-- quantity: integer (nullable = true)
 |-- discount: double (nullable = true)
 |-- profit: double (nullable = true)
 |-- order_year: integer (nullable = true)

+------+--------------+----------+----------+------------+-----------+---------------+---------+-

## What is done so far: 

1. read from bronze

2. clean collumn
remove null

3. deduplicate
key level / row level

create tables for duplicates 
4. write to silver
  add order year for partition
check silver talbe


ttd: 
check row number 

***
ttd databricks 

check bronze row  99

duplicate row 64/96

silver row number->   14/ 96

-> 288



?? 



NOTE: status in obsidien - apr 24